feat: adds sortable fields to studio content search index [FC-0059]

pomegranited opened the pull request:
<!--

Note: Please refer to the Support Development Guidelines on the wiki page to consider backporting to active releases:
https://openedx.atlassian.net/wiki/spaces/COMM/pages/4248436737/Support+Guidelines+for+active+releases

Please give your pull request a short but descriptive title.
Use conventional commits to separate and summarize commits logically:
https://open-edx-proposals.readthedocs.io/en/latest/oep-0051-bp-conventional-commits.html

Use this template as a guide. Omit sections that don't apply.
You may link to information rather than copy it, but only if the link is publicly readable.
If the linked information must be private (because it contains secrets), clearly label the link as private.

-->

## Description

This PR adds new datetime fields to the serialized library block metadata: `created`, `modified`, and `last_published`, and adds support for sorting search results by these dates or the block title/`display_name`. 

Also adds `last_published` to the list of fields we can filter by, so that e.g. sorting by "Recently Published" can also filter out unpublished content.

This change should not impact any current users, as  content libraries v2 are not yet ready for production use.

## Supporting information

Part of https://github.com/openedx/frontend-app-course-authoring/issues/1038

## Testing instructions

See https://github.com/openedx/frontend-app-course-authoring/pull/1147 for full testing instructions.

## Deadline

ASAP -- to meet project deadlines

## Other information

These changes can be deployed by running the management command:

```
python manage.py cms reindex_studio --experimental
```

pomegranited committed:
feat: adds sortable fields to studio content search index

* new datetime fields added to library component serializer: created,
  modified, and last_published
* new datetime fields are stored in Meilisearch as POSIX timestamps for
  sorting
* Meilisearch search ranking updated to allow sorting by display_name or
  these new datetime fields
* Updates tests to verify these new properties and client settings

openedx-webhooks commented:
Thanks for the pull request, @pomegranited!

### What's next?

*Please work through the following steps to get your changes ready for engineering review:*

#### :radio_button: Get product approval

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.

#### :radio_button: Provide context

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads





#### :radio_button: Get a green build

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.



#### :radio_button: Let us know that your PR is ready for review:

### Who will review my changes?


This repository is currently maintained by `@openedx/wg-maintenance-edx-platform`.  Tag them in a comment and let them know that your changes are ready for review.


### Where can I find more information?

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)

### When can I expect my changes to be merged?

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*

<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

pomegranited committed:
fix: reduce queries made when serializing library block metadata

* Instead of using `latest.modified` (which always queries the
  database), use `draft.latest`
* However, we have to increase the query count by 1 query per block,
  because the published_at date can only be found in the publish log,
  which must be fetched individually for each block.

pomegranited commented on openedx/core/djangoapps/content_tagging/tests/test_objecttag_export_helpers.py:
```diff
@@ -441,7 +441,7 @@ def test_build_library_object_tree(self) -> None:
         Test if we can export a library
         """
-        with self.assertNumQueries(8):
+        with self.assertNumQueries(11):
```
My [initial changes to `LibraryXBlockMetadata.from_component`](https://github.com/openedx/edx-platform/pull/35103/commits/cf9ab01300a8ba30ab037355af46be1bdf7597b1#diff-a760c3e87f2f671ac2453abbbaf6f10ca052c7acf1c46c1d36f24646e1d52a00R219) brought [this query count up to 14](https://github.com/openedx/edx-platform/actions/runs/9855392496/job/27210174606?pr=35103), because it used `component.versioning.latest`.

[Using `component.versioning.draft` instead brought the query count down to 11](https://github.com/openedx/edx-platform/pull/35103/commits/5f71cc3074dfd6024079b06a2279ab6208065a0d#diff-a760c3e87f2f671ac2453abbbaf6f10ca052c7acf1c46c1d36f24646e1d52a00R219), because the authoring API doesn't give us a more efficient way to fetch the `published_at` date than from the `last_publish_log`.

Welcome any advice on how to avoid the increased query count here!

bradenmacdonald replied:
This seems like an @ormsbee question if I'm correctly understanding that it's about efficiently fetching modified dates from learning core's `authoring` API.

yusuf-musleh commented on openedx/core/djangoapps/content/search/handlers.py:
```diff
@@ -96,6 +97,7 @@ def xblock_deleted_handler(**kwargs) -> None:
 
 
 @receiver(LIBRARY_BLOCK_CREATED)
+@receiver(LIBRARY_BLOCK_UPDATED)
 @only_if_meilisearch_enabled
 def library_block_updated_handler(**kwargs) -> None:
     """
```
When I create a new component in the library, it gets indexed correctly and it shows up in the mfe when I refresh. However when I update the block inside the library, the index doesn't get updated. I've looked into this a bit after our discussion, it seems like the `LIBRARY_BLOCK_UPDATED` is actually not being fired.

The only place where I found it in the codebase was here: 
https://github.com/openedx/edx-platform/blob/96da9902c60709ea5a46eb30d3782e4cad9d84b7/openedx/core/djangoapps/content_libraries/api.py#L724-L729

However the encompassing function `set_library_block_olx` isn't actually being called when I update a component inside my library. I think we're probably missing firing the `LIBRARY_BLOCK_UPDATED` in the right spot when updating the library component. I haven't looked into exactly where that would be, but thought I'd mention it in case it helps in narrowing down the issue.

pomegranited replied:
@yusuf-musleh Oddly, I don't think we have an API method for updating library block metadata yet. There definitely [isn't support for editing library blocks the frontend](https://github.com/openedx/frontend-app-course-authoring/issues/1040#issuecomment-2135814907) yet either, so I guess that adding this will be part of that scope?

How did you update your library component?

yusuf-musleh replied:
@pomegranited Oh I see, I've been using the library authoring mfe to add/edit components inside the libraries, not the course authoring mfe.

The endpoint that the library authoring mfe is calling when updating the library components is [here](https://github.com/openedx/edx-platform/blob/57fd67158cd19294a4366ce09b7573657c8ae97d/openedx/core/djangoapps/xblock/rest_api/views.py#L211). I'm not sure if it is out of scope for this task to add the `LIBRARY_BLOCK_UPDATED` trigger to this endpoint or if that is even something we want to do in the first place, since we are migrating everything to the course authoring mfe and the endpoints called might likely change.

pomegranited replied:
Ah excellent, thank you for locating that endpoint! I'm pretty sure the studio index wants to know about these updates, and it's a small change, so I can do it here.

I'm not sure whether the xblock REST API is used for editing course blocks too.. so will detect whether the XBlock is part of a library or a course, and send `LIBRARY_BLOCK_UPDATED` or `XBLOCK_UPDATED` as appropriate.

pomegranited replied:
cf [b7f8f8e](https://github.com/openedx/edx-platform/pull/35103/commits/b7f8f8e69916a11420de7bdfaea6f88258be2710)

yusuf-musleh reviewed (commented):
@pomegranited Great work on this! I've tested it out with the frontend and it's working quite well. I just left a comment the issue with updating the index when updating the component in the library. I shared an idea, hopefully it might help solve it.

pomegranited committed:
Merge branch 'master' into jill/index-component-dates

pomegranited committed:
feat: raise LIBRARY_BLOCK_UPDATED in the xblock edit REST API

rpenido reviewed (approved):
LGTM :+1:
Thank you for your work @pomegranited!

- [x] I tested this using the instructions from https://github.com/openedx/frontend-app-course-authoring/pull/1147
- [x] I read through the code
- [x] ~I checked for accessibility issues~
- [x] Includes documentation

**Note:** While testing this, I found an issue (not related) updating my index. I create a PR to circumvent this:
- https://github.com/openedx/edx-platform/pull/35139

pomegranited committed:
Merge branch 'master' into jill/index-component-dates

pomegranited committed:
fix: allow content search to filter on last_modified

pomegranited committed:
Merge remote-tracking branch 'origin/master' into jill/index-component-dates

pomegranited committed:
Merge remote-tracking branch 'origin/master' into jill/index-component-dates

bradenmacdonald commented on openedx/core/djangoapps/xblock/rest_api/views.py:
```diff
@@ -254,6 +258,20 @@ def post(self, request, usage_key_str):
+            )
+
+        except InvalidKeyError:
+            # TODO raise XBLOCK_UPDATED if this API is ever used for course blocks.
```
The XBlock API isn't supposed to know anything about specific contexts, so instead of taking an "if library, do X, else if course do Y" approach, the better option is something like this:

```python
context_impl = get_learning_context_impl(usage_key)
context_impl.send_updated_event(block)
```

Then add `send_updated_event` to both the abstract [`class LearningContext`](https://github.com/openedx/edx-platform/blob/master/openedx/core/djangoapps/xblock/learning_context/learning_context.py) and the [`LibraryContext`](https://github.com/openedx/edx-platform/blob/master/openedx/core/djangoapps/content_libraries/library_context.py) subclass. Since that's fairly simple, would you mind doing it now? I know we discussed that this approach would be fine but I am hoping this is a small change.

pomegranited replied:
@bradenmacdonald that's much nicer, thank you! See [cc67395](https://github.com/openedx/edx-platform/pull/35103/commits/cc6739581ee7fc54aa78c13127dad291dfaec1c1)

bradenmacdonald replied:
Looks nice! Just one issue with a line that's not needed, then I'll approve this :)

pomegranited committed:
refactor: use LearningContext to send "block updated" event

pomegranited committed:
Merge branch 'master' into jill/index-component-dates

bradenmacdonald commented on openedx/core/djangoapps/xblock/rest_api/views.py:
```diff
@@ -258,19 +255,10 @@ def post(self, request, usage_key_str):
-            # TODO raise XBLOCK_UPDATED if this API is ever used for course blocks.
-            pass
+        # Signal that we've modified this block
+        usage_key = UsageKey.from_string(usage_key_str)
```
I think usage_key is already defined earlier in this function scope, so this line isn't needed.

pomegranited replied:
Oops! Sorry.. fixed with [ceaf314](https://github.com/openedx/edx-platform/pull/35103/commits/ceaf3142d5599796b87505ac418d69283fd9fe68)

pomegranited committed:
fix: nit, don't redeclare usage_key

openedx-webhooks commented:
@pomegranited 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/edx-platform&entry.867055334=https://github.com/openedx/edx-platform/pull/35103&entry.1484655318=https://github.com/pomegranited&entry.752974735=2024-07-09+10:28&entry.1917517419=2024-07-25+18:18&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->

pomegranited commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -203,13 +206,18 @@ def from_component(cls, library_key, component):
             display_name=component.versioning.draft.title,
+            created=component.created,
+            modified=component.versioning.draft.created,
+            last_published=None if last_publish_log is None else last_publish_log.published_at,
```
Hi @ormsbee 👋 
I added this code to record the published date for library components in the Studio search index. However, it doesn't work, seemingly because components don't get publish records recorded when they are published -- only the parent library does.

However, there isn't a date available on the `component.versioning.published` ComponentVersion -- the `published.created` matches the `draft.modified` date after publishing, and doesn't represent the actual date published.

What would be the best way to address this issue?

CC @bradenmacdonald

pomegranited replied:
cf https://github.com/openedx/edx-platform/pull/35195

pomegranited replied:
And https://github.com/openedx/openedx-learning/pull/204

ormsbee replied:
I don't remember if there's a good helper for it at the moment, but I think you can get the datetime when a Component was published by following these relations:

Component.publishable_entity -> PublishableEntity.published -> Published.publish_log_record -> PublishLogRecord.publish_log -> PublishLog. published_at

bradenmacdonald commented:
@pomegranited I think one of the new tests might be flaky: https://github.com/openedx/frontend-app-course-authoring/actions/runs/10166343467/job/28116281335

Might need to wrap the failing line in a `waitFor` ?
